// express-mung/.eslintrc.js
module.exports = {
    plugins: ['mocha'],
    extends: ["standard", "plugin:mocha/recommended"],
    globals: {
      afterAll: false,
      afterEach: false,
      beforeAll: false,
      beforeEach: false,
      context: false,
      describe: false,
      expect: false,
      it: false,
      require: false,
      setup: false,
      specify: false,
      suite: false,
      suiteSetup: false,
      suiteTeardown: false,
      teardown: false,
      test: false,
    },
    rules: {
      "mocha/handle-done-callback": ["error"],
      "mocha/max-top-level-suites": ["error"],
      "mocha/no-exclusive-tests": ["error"],
      "mocha/no-global-tests": ["error"],
      "mocha/no-hooks": ["error"],
      "mocha/no-hooks-for-single-case": ["error"],
      "mocha/no-identical-title": ["error"],
      "mocha/no-mocha-arrows": ["error"],
      "mocha/no-nested-tests": ["error"],
      "mocha/no-pending-tests": ["error"],
      "mocha/no-return-and-callback": ["error"],
      "mocha/no-setup-in-describe": ["error"],
      "mocha/no-sibling-hooks": ["error"],
      "mocha/no-skipped-tests": ["error"],
      "mocha/no-synchronous-tests": ["error"],
      "mocha/no-top-level-hooks": ["error"],
      "mocha/prefer-arrow-callback": ["error"],
      "mocha/valid-suite-description": ["error"],
      "mocha/valid-test-description": ["error"],
      "mocha/no-async-describe": ["error"],
      "indent": ["error", 4],
      "keyword-spacing": ["error", {
        "before": true, "after": true,
      }],
      "space-in-parens": ["off"],
      "space-before-function-paren": ["off"],
      "comma-dangle": ["error", {
        "arrays": "only-multiline",
        "objects": "only-multiline",
        "imports": "only-multiline",
        "exports": "only-multiline",
        "functions": "ignore"
      }],
    }
};
